### PR TITLE
reworked implementation of Singularity integration

### PR DESCRIPTION
**Pull request by boegel**
@shahzebsiddiqui This is a thorough revision of your work in https://github.com/easybuilders/easybuild-framework/pull/2332, which should make it good to merge into `develop` for inclusion in EasyBuild v3.6.0.

Short overview of changes:

* more general command line options: `--container*`
  * `--containerize` or `-C` to enable generating of container recipe
    * (replaces `--singularity`)
  * `--containerpath`: location to store generated container recipes & images
    * (replaces `--singularitypath`)
  * `--container-build-image`: enable building of container image (requires `sudo`)
    * (replaces `--build-image`)
  * `--container-base` to specify Singularity bootstrap mechanism (`docker`, `localimage` or `shub`)
    * (replaces `--singularity-bootstrap`)
  * `--container-image-format`: `ext3`, `sandbox`, `squashfs` (default)
    * (replaces `--imageformat`)
  * `--container-image-name` (optional): name for container recipe/image (derived from 1st easyconfig file if not specified)
    * (replaces `--imagename`)
  * `--container-type`: currently only accepts `docker` and `singularity`, only `singularity` is actually implemented
    * (replaces for `--type`, which was clearly way too generic)

* `easybuild.tools.container` module replaces `easybuild.tools.singularity` package
* significantly cleaned up implementation
   * avoid duplicate code
   * split code into more functions (easier to digest)
   * leverage existing functions like `run_cmd` and `det_full_ec_version`
   * use template for container recipe
* introduces additional constants where it makes sense, to avoid hardcoding & "magic strings"
* added dedicated tests for container support (see `test/framework/container.py`)

cc @FooBarQuaxx

### CODE REVIEW

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [inline easybuild/tools/containers.py]: Can you try running the following but with your new options 
```
eb OpenMPI-2.1.0-GCC-6.3.0-2.28.eb --singularity --singularity-bootstrap shub:shahzebsiddiqui/eb-singularity:centos-7.3.1611
```

to confirm os packages in %post are installed as follows

```
yum install -y libibverbs-dev || true
yum install -y libibverbs-devel || true
```

If package is not available then singularity will stop immediately during yum step. With the conditional it would proceed even if package X is not available. 

Idea is to let singularity build image even if package X is not available so this section can be expanded where users can install arbitrary packages in the future. If user request package X and is not available we need to decide whether we exit hard or ignore package.

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [inline easybuild/tools/containers.py:240]: @boegel

it may be a good idea when user is overwriting container with --force to remove existing container before building new container. Also note that ``singularity build --force`` is also an option but that is something I have not played around with. 

If this option is useful in future we would need to differentiate between ``eb --force`` and ``singularity build --force`` to avoid confusion. 


```
    -f|-F|--force   Force a rebootstrap of a base OS (note: this does not
                    delete what is currently in the image, just causes the core
                    to be reinstalled)
```

I am guessing singularity build -F may be useful if you want to build same container and there is update in the bootstrap image.

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [inline easybuild/tools/containers.py:141]: @boegel

Should we also allow bootstrap from sandbox container. Sandbox container is just a directory so it is hard to check if a directory is a singularity container or just a regular directory. Something to think about for future updates to eb-singularity support.

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: With the enhanced implementation, it now looks like this:

```
$ grep %post -A 1 Singularity.OpenMPI-2.1.0-GCC-6.3.0-2.28
%post
yum install -y libibverbs-dev || yum install -y libibverbs-devel || yum install -y rdma-core-devel
```

Which is more correct than what you had imho, i.e. if *one* of these is installed, it's fine. 

With your approach, it was possible multiple ones got installed, at least in theory.

Comment (FooBarQuaxx) [reply easybuild/tools/containers.py]: you can also `yum install --skip-broken install -y libibverbs-dev libibverbs-devel rdma-core-devel || true`. The final `true` is needed because it can be the case that all of the dependencies are not available. I don't know whether that's desirable or not though.

File: easybuild/tools/containers.py
Comment (FooBarQuaxx) [inline easybuild/tools/containers.py:56]: It makes sense to install the same version as the one you creating the recipe from.

File: easybuild/main.py
Comment (FooBarQuaxx) [inline easybuild/main.py:392]: I think you should put this after resolving dependencies so you will be able to extract the os dependencies for all the applications that should be build and not only the user requested one.

File: easybuild/tools/containers.py
Comment (FooBarQuaxx) [inline easybuild/tools/containers.py]: I think you should return very early from this function in case no container build is requested.

File: easybuild/tools/containers.py
Comment (FooBarQuaxx) [inline easybuild/tools/containers.py]: This is not enough as noted previously, I think you should not assume that only one build needs to have its os dependencies provided.

File: easybuild/tools/containers.py
Comment (FooBarQuaxx) [inline easybuild/tools/containers.py]: Maybe you should check for sudo existence while checking for singularity image build tools presence.

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [reply easybuild/tools/containers.py]: i agree with @FooBarQuaxx should add ``|| true`` at the end in case none of the deps are not available.

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [reply easybuild/tools/containers.py]: I agree with @FooBarQuaxx easybuild will need to add all the osdeps from the dependency list ``--robot`` which could imply searching all easyconfigs. I'd recommend you create a set data structure and add os dependency and then iterate over the set and install the pkgs. 

When easyconfigs specify os dependency as ``osdependencies = [('libibverbs-dev', 'libibverbs-devel', 'rdma-core-devel')] `` then this could be processed as each element and added to the set. It would be okay to add libibverbs-dev and libibverbs-devel into the system since we know one of these packages are not going to be found. In the future it would be good to tag os dependency with the linux distribution so you can just pick the os dependency based on the bootstrap os in the container.

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [inline easybuild/tools/containers.py:99]: @boegel So a first time user trying the eb-singularity stuff would have no clue what base container to use.  

```
[siddis14@amrndhl1157 easybuild-framework]$ eb M4-1.4.17.eb -C --experimental
== temporary log file in case of crash /tmp/eb-LxMMKc/easybuild-LHIN2Y.log
== Singularity tool found at /usr/local/bin/singularity
== Singularity version '2.4' is 2.4 or higher ... OK
ERROR: --container-base must be specified
```

Either add ``--container-base shub:shahzebsiddiqui/eb-singularity:centos-7.3.1611`` and ``--container-base shub:shahzebsiddiqui/eb-singularity:centos-7.4.1708`` in the Error message or in the help section. Another option is if user don't specify ``--container-base`` easybuild can try to use one of the base containers either centos 7.3 or 7.4. In the future, I would like to move the base container into easybuilders repo instead of mine. But for now users have to use my container image since that is the only image available in SHUB.

**Review by shahzebsiddiqui — CHANGES_REQUESTED** [review]
@boegel Please address this issue, it would be important for new users to understand what base-container to use.

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [reply easybuild/tools/containers.py]: @boegel Also running into an issue when issuing singularity command. It is showing the password and hanging. The issue is with `run_cmd` function

I have X out my password but it shows up on the console and after pressing ENTER it just waits. Also sudo password should be invoked once and any subsequent builds will work without password. That is how it worked in the past. 

```
[siddis14@amrndhl1157 easybuild-framework]$ eb M4-1.4.17.eb -C --experimental --container-base shub:shahzebsiddiqui/eb-singularity:centos-7.3.1611 --force --container-build
== temporary log file in case of crash /tmp/eb-i5jViB/easybuild-r745VY.log
== Singularity tool found at /usr/local/bin/singularity
== Singularity version '2.4' is 2.4 or higher ... OK
== WARNING: overwriting existing container recipe at /home/siddis14/.local/easybuild/containers/Singularity.M4-1.4.17 due to --force
== Singularity definition file created at /home/siddis14/.local/easybuild/containers/Singularity.M4-1.4.17
== Running 'sudo singularity build  /home/siddis14/.local/easybuild/containers/M4-1.4.17.simg /home/siddis14/.local/easybuild/containers/Singularity.M4-1.4.17', you may need to enter your 'sudo' password...
XXXXXX
```

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: Thanks for the suggestion @FooBarQuaxx, that's indeed cleaner. :)

In that case we're pretty sure that the `eb` installation will not work, so why not fail early?

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py:141]: Well, the easy thing to do would be to just remove the extension check all together, and trust that the users knows what he/she is doing... Less code to worry about too ;)

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py:56]: I guess this is up for debate, and part of the reason why we're putting this under `--experimental` for now.

The worry was that if we don't update EasyBuild, you'll be stuck to a particular EasyBuild version in this container, which can be a problem if you'll be using the container image you create as a base for others.

I guess we could consider making this optional, and having some other mechanism (like allowing to specify a particular EasyBuild version to use in the container).

For now, I'd leave this as is, we can revisit this later?

File: easybuild/main.py
Comment (boegel) [reply easybuild/main.py:392]: Ah, that's indeed a good point...

But then we'll need to discriminate between the easyconfigs that were specified initially, and the full set of easyconfigs after resolving dependencies.

OK if we leave this for later enhancement? Maybe open an issue on this at https://github.com/easybuilders/easybuild-framework/issues so we won't forget?

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: Very good point, thanks, I'll change this so that it only runs when `--container-build-image` is used.

File: easybuild/main.py
Comment (FooBarQuaxx) [reply easybuild/main.py:392]: I was thinking about just using the passed arguments/options to the `eb` inside the recipe but only removing the container specific options.

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [reply easybuild/tools/containers.py:56]: @boegel  i agree with you having the `pip install -U easybuild` in the container recipe is better than having to update or support many easybuild base container. It would be overkill to create a base container for every easybuild version but rather let pip install the latest version inside the container so hopefully the eb on the host and inside the container are the same version.

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: For now, I'll change this to consider OS deps of all easyconfigs that are passed down to here.

I would like to leave also taking into account OS dependencies after resolving dependencies for future enhancement (you'll be able to get around this by specifying the easyconfig on the `eb` command for the time being).

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [reply easybuild/tools/containers.py:141]: Let me ask with singularity if bootstraping with sandbox is a good idea, as for the logic I think having a check for local image is good. Ideally we need a singularity python api to identify something is a valid singularity container from shub, docker and sandbox container.

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py:99]: How about changing this to `ERROR: --container-base must be specified (see 'eb --help')`, or pointing to the documentation (URL to be figured out)?

since:

```
$ eb --help | grep container-base
    --container-base=BASE
                        --container-base localimage:/path/to/image.img,
                        --container-base shub:<image>:<tag>, --container-base
```

File: easybuild/tools/containers.py
Comment (FooBarQuaxx) [reply easybuild/tools/containers.py]: That was the thing, I don't know which approach makes more sense at that point. But I guess carrying on is kind of more optimistic.

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: @FooBarQuaxx Makes sense, I'll check for `sudo` too.

@shahzebsiddiqui I'm not seeing that issue (on `macOS`, with a fake `singularity`), I get a proper password prompt, which repeats if I mistype the password...

```
== Running 'sudo singularity build  /Users/kehoste/.local/easybuild/containers/cURL-7.58.0-GCCcore-6.4.0.simg /Users/kehoste/.local/easybuild/containers/Singularity.cURL-7.58.0-GCCcore-6.4.0', you may need to enter your 'sudo' password...
Password:
Password:
== Singularity image created at /Users/kehoste/.local/easybuild/containers/cURL-7.58.0-GCCcore-6.4.0.simg
== Temporary log file(s) /tmp/eb-5596mm/easybuild-L8YGho.log* have been removed.
== Temporary directory /tmp/eb-5596mm has been removed.
```

Can you try again with `eb -ld` to see where it hangs?

File: easybuild/main.py
Comment (boegel) [reply easybuild/main.py:392]: That makes sense, that's also what we do with `--job` (except that there we also remove `--robot` since each job takes care of one installation and dependencies are set between them), see https://github.com/easybuilders/easybuild-framework/blob/master/easybuild/tools/parallelbuild.py#L129

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: done in 53ad538f21ed2c332726314a86e9741997d2a800

File: easybuild/main.py
Comment (boegel) [reply easybuild/main.py:392]: Follow-up in https://github.com/easybuilders/easybuild-framework/issues/2475

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: @shahzebsiddiqui Note that it's possible that `sudo` doesn't ask your password if you've recently entered it, so if you see a 'hang' it's probably because `singularity` is running...

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [inline easybuild/tools/containers.py]: @boegel 

the `run_cmd` does not show streaming output of ``singularity build``. It is misleading and looks as if system is waiting for user input but sometimes it is actually building the container but no output is shown while other times it will ask for sudo password  which happens after a set interval.

Users have to reverse engineer to find logs for troubleshooting, if the container build fails there would be no log since the logs are written inside the container. So it is important the output is shown on console before we merge this. 

```
[siddis14@amrndhl1157 easybuild-framework]$ eb M4-1.4.17.eb -C --experimental --container-base shub:shahzebsiddiqui/eb-singularity:centos-7.3.1611 --force --container-build
== temporary log file in case of crash /tmp/eb-EHG6Ck/easybuild-rmBXp3.log
== Singularity tool found at /usr/local/bin/singularity
== Singularity version '2.4' is 2.4 or higher ... OK
== WARNING: overwriting existing container recipe at /home/siddis14/.local/easybuild/containers/Singularity.M4-1.4.17 due to --force
== Singularity definition file created at /home/siddis14/.local/easybuild/containers/Singularity.M4-1.4.17
== WARNING: overwriting existing container image at /home/siddis14/.local/easybuild/containers/M4-1.4.17.simg due to --force
== Running 'sudo singularity build  /home/siddis14/.local/easybuild/containers/M4-1.4.17.simg /home/siddis14/.local/easybuild/containers/Singularity.M4-1.4.17', you may need to enter your 'sudo' password...
== Singularity image created at /home/siddis14/.local/easybuild/containers/M4-1.4.17.simg
== Temporary log file(s) /tmp/eb-EHG6Ck/easybuild-rmBXp3.log* have been removed.
```

**Review by shahzebsiddiqui — CHANGES_REQUESTED** [review]

File: easybuild/tools/config.py
Comment (shahzebsiddiqui) [inline easybuild/tools/config.py:522]: @boegel 

it appears that containerpath is not modified when using configuration file

```
[siddis14@amrndhl1157 easybuild-framework]$ cat ~/.local/easybuild/config.cfg
[config]
containerpath = /tmp
[siddis14@amrndhl1157 easybuild-framework]$ eb --show-full-config | grep containerpath
containerpath                             (D) = /home/siddis14/.local/easybuild/containers

```

It is working via command line and environment variable

```
[siddis14@amrndhl1157 easybuild-framework]$ export EASYBUILD_CONTAINERPATH=/tmp
[siddis14@amrndhl1157 easybuild-framework]$ eb --show-full-config | grep containerpath
containerpath                             (E) = /tmp
```

This feature was working before. Can you please look into this.

**Review by shahzebsiddiqui — CHANGES_REQUESTED** [review]
containerpath is not modified when using configuration file

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [inline easybuild/tools/containers.py:137]: @boegel 

Running into some weird issues with localimage 

for instance if you have a space character between ":" and the container path I get the following

```
[siddis14@amrndhl1157 easybuild-framework]$ eb Anaconda3-4.2.0.eb -C --container-base localimage: ~/.local/easybuild/containers/git-lfs-1.1.1/ --experimental
== temporary log file in case of crash /tmp/eb-iXLynP/easybuild-ScZ5Ui.log
ERROR: Singularity base image at specified path does not exist:
```

Shell expansion does not work

```
[siddis14@amrndhl1157 easybuild-framework]$ eb Anaconda3-4.2.0.eb -C --container-base localimage:~/.local/easybuild/containers/git-lfs-1.1.1/ --experimental
== temporary log file in case of crash /tmp/eb-MrrU23/easybuild-ht00Iw.log
ERROR: Singularity base image at specified path does not exist: ~/.local/easybuild/containers/git-lfs-1.1.1/
```

This works

```
[siddis14@amrndhl1157 easybuild-framework]$ eb Anaconda3-4.2.0.eb -C --container-base localimage:$HOME/.local/easybuild/containers/git-lfs-1.1.1/ --experimental
== temporary log file in case of crash /tmp/eb-b2rOE1/easybuild-pqXfed.log
ERROR: Invalid image extension '' must be .img or .simg
```



I think we can trim any space and newline characters before processing the argument. I think this should be done in parse_container_base method 

If you can get the shell expansion working that would be nice if not we can create an issue to track this for future.

**Review by shahzebsiddiqui — CHANGES_REQUESTED** [review]
issue with shell expansion and argument passing for localimage

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [reply easybuild/tools/containers.py:99]: I will document bootstrap easybuild containers for shub in the easybuild documentation. But users don't want to refer to documentation everytime to remember the SHUB easybuild base container since they will be using it.

It would be nice either to add the message in either eb --help or in the error message something like 

``` 
ERROR: --container-base must be specified. See below for examples

eb --container-base localimage:$HOME/.local/easybuild/containers/M4-1.4.17.simg
eb --container-base shub:shahzebsiddiqui/eb-singularity:centos-7.3.1611
eb --container-base shub:shahzebsiddiqui/eb-singularity:centos-7.4.1708
eb --container-base docker:centos:7.3.1611
```

I'll let you decide on the messaging.

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [inline easybuild/tools/containers.py]: This is a typo it should be ``--writable`` and not ``--writeable``

I get the following error

```
[siddis14@amrndhl1157 easybuild-framework]$ eb M4-1.4.18.eb -C --container-base shub:shahzebsiddiqui/eb-singularity:centos-7.3.1611 --container-build --container-image-format=ext3 --experimental --force
== temporary log file in case of crash /tmp/eb-l2lS1C/easybuild-rMxfEQ.log
== Singularity tool found at /usr/local/bin/singularity
== Singularity version '2.4' is 2.4 or higher ... OK
== Singularity definition file created at /home/siddis14/.local/easybuild/containers/Singularity.M4-1.4.18
== Running 'sudo singularity build --writeable /home/siddis14/.local/easybuild/containers/M4-1.4.18.img /home/siddis14/.local/easybuild/containers/Singularity.M4-1.4.18', you may need to enter your 'sudo' password...
[sudo] password for siddis14:
ERROR: cmd "sudo singularity build --writeable /home/siddis14/.local/easybuild/containers/M4-1.4.18.img /home/siddis14/.local/easybuild/containers/Singularity.M4-1.4.18" exited with exit code 1 and output:
ERROR: Unknown option: --writeable
Cleaning up...
```

See documentation 

```

[siddis14@amrndhl1157 easybuild-framework]$ singularity build --help | grep writable
        writable:   Legacy writable image format
                    it is legacy or writable format)
    -w|--writable   Build a writable image (warning: deprecated due to sparse
        $ singularity exec --writable /tmp/debian apt-get install python

```

**Review by shahzebsiddiqui — CHANGES_REQUESTED** [review]
error building ext3 container format, typo in singularity build command

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: I've added support to `run_cmd` for enabling streaming output in https://github.com/easybuilders/easybuild-framework/pull/2476.

Once that PR gets merged, I'll update this PR to include it, and enable it for `run_cmd("sudo singularity ...")`

File: easybuild/tools/config.py
Comment (boegel) [reply easybuild/tools/config.py:522]: @shahzebsiddiqui The default configuration file is in `$HOME/.config/easybuild/config.cfg`, not `$HOME/.local/easybuild/config.cfg`, see http://easybuild.readthedocs.io/en/latest/Configuration.html#default-configuration-files and the output of `eb --show-default-configfiles`

works fine if you use the right location:

```
$ cat ~/.config/easybuild/config.cfg
[config]
containerpath = /tmp

$ eb --show-config | grep containerpath
containerpath         (F) = /tmp
```

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py:137]: This is not so easy to fix... If you use a space, `eb` will take `localimage` as an argument to `--container-base`, and the specified path of the image as if it were an easyconfig file, and it will no longer be related to `--container-base` at all...
All arguments on the `eb` command that are not tied to an option are assumed to be easyconfig files.

So, this is not so trivial to fix, unless we introduce another option like `--container-base-args` or something like that:

```
eb Anaconda3-4.2.0.eb -C --container-base localimage --container-base-args $HOME/.local/easybuild/containers/git-lfs-1.1.1/ --experimental
```

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: woops, fix pushed

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py:99]: `eb --help` already includes useful suggestions imho:

```
eb --help | grep container-base -A 1
    --container-base=BASE
                        Base for container image. Examples (for Singularity):
--
                        --container-base localimage:/path/to/image.img,
                        --container-base shub:<image>:<tag>, --container-base
                        docker:<image>:<tag>  (type <type 'str'>)
```

File: easybuild/tools/containers.py
Comment (FooBarQuaxx) [inline easybuild/tools/containers.py]: This constant and the one on the following line are duplicates of the one on `easybuild.tools.config`. Any reason for that?

File: easybuild/tools/containers.py
Comment (FooBarQuaxx) [inline easybuild/tools/containers.py]: You probably have to add `-y` options.

File: easybuild/tools/containers.py
Comment (FooBarQuaxx) [inline easybuild/tools/containers.py]: What about `img_name = def_file.split('.', 1)[1]`

File: easybuild/tools/containers.py
Comment (FooBarQuaxx) [inline easybuild/tools/containers.py]: Is this necessary? Correct me if I'm wrong but I think the module creation step of the build takes care of loading the depended on modules.

File: easybuild/tools/config.py
Comment (shahzebsiddiqui) [reply easybuild/tools/config.py:522]: You are right, thanks for the clarification. I got the paths mixed up. This is working.

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [reply easybuild/tools/containers.py:137]: @boegel no worries let's hold off on this if it is too much work.

Is the shell expansion something that can be fixed easy so if user does the following ``--container-base localimage:~/some.img`` it will resolve properly to the $HOME directory instead of throwing an error.

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: @FooBarQuaxx This is not related to dependencies, we only construct a list of module names for the easyconfigs that were specified on the command line here.

We can simplify this to use `e['ec'].full_mod_name` though, so I'll make that change.

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: No, I'll fix that, these constants shouldn't be redefined, indeed...

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: Nice catch, thanks, will fix

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py]: That's better indeed, will change.

File: easybuild/tools/containers.py
Comment (boegel) [reply easybuild/tools/containers.py:137]: The tab completion or shell expansion of ~ won't work because of the preceding `:`, we can't easily fix that unless we go with a separate option like `--container-base-args`

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [reply easybuild/tools/containers.py]: i agree with @FooBarQuaxx eb just needs to load the module for the calling application and eb will load all the dependent modules. This is true for Flat Naming Scheme since all modules are present in one moduletree. 
I think we can remove the logic and references for variable ``mod_names``

Comment (boegel) [reply easybuild/tools/containers.py]: @shahzebsiddiqui No, how will we know which modules to load in the `%environment` section of the Singularity image? We need to module names for this...

File: easybuild/tools/containers.py
Comment (shahzebsiddiqui) [reply easybuild/tools/containers.py:137]: okay let's leave it for now and remember this for future.

**Review by shahzebsiddiqui — APPROVED** [review]
I think all the issues have been addressed for the 1st release of eb-singularity. Some issues that we found will be addressed subsequent eb release and these issues are reported in https://github.com/easybuilders/easybuild-framework/issues

**Pull request merged**
